### ISSUE TITLE
[Feature] Extended skin metadata/settings/permissions

### ISSUE THREAD

**Issue by Drewol** [Drewol/unnamed-sdvx-clone#506]
Opening for feedback from skin authors and contributors.

I have been thinking about a new feature for skins to better manage versions and skin settings.

The reason for introducing permissions is to make skins more secure and for users to be more aware of what skins can actually do on their computer. At the moment skins are given the full set of lua's built in libraries which includes a lot of file and io access that most skins don't actually need.

Another thing is something like a "target game version" so that the game can know if a skin targets an older version or a newer version of the game if the user hasn't updated for some reason.

The file could look something like this for example:

skininfo.json
```jsonc
{
    "name": "Default",
    "author": "Drewol",
    "version": [0, 1, 0],
    "game_version": [0, 5, 1],
    "config_definitions": [], //Like the current "config-definitions.json" but better
    "permissions": ["file", "sys", "http"], //TBD
    "website": "https://github.com/Drewol/unnamed-sdvx-clone"
}
```

In addition to these fields, I would also prefer to have things such as what's proposed in #505 in here instead of something that may or may not be called by lua.

**Comment by FajsiEx** [comment]
I completely agree with limiting the skin's permissions, unless the user gives informed consent. I feel like there could be (if possible) a default permission for the skin to have filesystem access to files in it's own skin folder (although even that shouldn't be needed if there are easier ways to achieve the same thing)

Target game version sounds really nice, as there is no way to check the currently installed USC version from the skin to display a message for example.

`version` field is also a nice addition since every skin handles it in a different way, and I think it's for the better to unify it. Would the skin version be available from this file in the skin code to do stuff like update checking like nightfall does? Or will skin updating be handled through USC itself?

Lastly, I support the addition of laser, track and other texture's default paths to this skininfo.json file. As for changing the laser colors the file could have `textures` field with something like this:

```json
{
    "textures": {
        "laser_l": {
            "default": "textures/lasers/default.png",
            "variants": [
                {
                    "name": "Blue",
                    "path": "textures/lasers/blue.png"
                }
            ]
        }
    }
}
```
just throwing ideas out there...

Overall excited for the changes! Good work :)

**Comment by Drewol** [comment]
@Hoshikara  The permissions are mostly to inform users but one approach would be to have them able to deny just some skin permissions for a skin and then the skin could have something like

```lua
if game.has_permission("http") then
    --get something from the web
end
```

@FajsiEx Currently I would still make it so skins have to manage the version checking/updating on their own so the version number in the new json would mainly be something for USC to load and display without having to load the entire skin.

**Comment by itszn** [comment]
Sounds good! I previously played with doing something like this when I was looking at reducing lua's permissions [like we do for background scripts](https://github.com/Drewol/unnamed-sdvx-clone/blob/master/Main/src/Background.cpp#L122).

A few other possible permissions: 
`all_file`: I think by default we can allow read from game dirs and write to the skin directory. So this would allow you to access anything.
`exec`: We might be able to keep most of os by default if we remove the functions that let you run subprocesses.
`dll`: Allows it to load arbitrary dlls. You would pretty much never need this unless someone actually uses live2d.

There are a few other lua libraries (coroutine, debug) which I don't really trust (like it might be possible to use them to escape out) but they also might be safe.

Might not hurt to check if there is any weird lua features current skins are using.

**Comment by Drewol** [comment]
I've written a draft for a schema and a translation of the default skin's config-definitions 98a848fba387b24696171879f5bdac17d0685726

Schema could probably be improved with how it handles config entry types, but it already provides pretty good autocompletes and validation
